modules/table/src/test/java/org/apache/ignite/internal/table/distributed/replication/PartitionReplicaListenerTest.java

```diff
@@ -1693,8 +1743,107 @@ private static Stream<Arguments> multiRowsWriteRequestTypes() {
+        when(tableVersion2.tableVersion()).thenReturn(NEXT_SCHEMA_VERSION);
+
+        when(catalogTables.table(tblId, txBeginTs.longValue())).thenReturn(tableVersion1);
+        when(catalogTables.table(eq(tblId), gt(txBeginTs.longValue()))).thenReturn(tableVersion2);
+
+        CompletableFuture<?> future = listenerInvocation.invoke(txId, key);
+
+        TransactionException ex = assertWillThrowFast(future, TransactionException.class);
+        assertThat(ex.code(), is(Transactions.TX_INCOMPATIBLE_SCHEMA_ERR));
+        assertThat(
+                ex.getMessage(),
+                is("Table schema was updated since the transaction was started [table=1, startSchema=1, operationSchema=2")
+        );
+    }
+
+    @CartesianTest
+    @CartesianTest.MethodFactory("multiRowsReadOrWriteRequestTypesFactory")
+    void multiRowRwReadsAndWritesFailIfTableSchemaVersionIncreasedSinceTxStart(
+            RequestType requestType, boolean onExistingRow, boolean full
+    ) {
+        testRwReadsAndWritesFailIfTableSchemaVersionIncreasedSinceTxStart(onExistingRow, (targetTxId, key) -> {
+            return doMultiRowRequest(targetTxId, List.of(marshalKeyOrKeyValue(requestType, key)), requestType, full);
+        });
+    }
+
+    @SuppressWarnings("unused")
+    private static ArgumentSets multiRowsReadOrWriteRequestTypesFactory() {
+        return ArgumentSets.argumentsForFirstParameter(multiRowsReadOrWriteRequestTypes())
+                .argumentsForNextParameter(false, true)
+                .argumentsForNextParameter(false, true);
+    }
+
+    private static Stream<RequestType> multiRowsReadOrWriteRequestTypes() {
+        return Arrays.stream(RequestType.values())
+                .filter(type -> RequestTypes.isMultipleRowsRwRead(type) || RequestTypes.isMultipleRowsWrite(type));
+    }
+
+    @CartesianTest
+    void replaceRequestFailsIfTableSchemaVersionIncreasedSinceTxStart(
+            @Values(booleans = {false, true}) boolean onExistingRow,
+            @Values(booleans = {false, true}) boolean full
+    ) {
+        testRwReadsAndWritesFailIfTableSchemaVersionIncreasedSinceTxStart(onExistingRow, (targetTxId, key) -> {
+            return doReplaceRequest(
+                    targetTxId,
+                    marshalKeyOrKeyValue(RequestType.RW_REPLACE, key),
+                    marshalKeyOrKeyValue(RequestType.RW_REPLACE, key),
+                    full
+            );
+        });
+    }
+
+    @CartesianTest
+    void rwScanRequestFailsIfTableSchemaVersionIncreasedSinceTxStart(
+            @Values(booleans = {false, true}) boolean onExistingRow,
+            @Values(booleans = {false, true}) boolean full
+    ) {
+        testRwReadsAndWritesFailIfTableSchemaVersionIncreasedSinceTxStart(onExistingRow, (targetTxId, key) -> {
+            return doRwFullScanRetrieveBatchRequest(targetTxId, full);
+        });
+    }
+
+    private UUID beginTx() {
+        return transactionIdFor(clock.now());
```

Comment:
So it's not a "begin", it just generates new TX ID. It confused be for a little bit